design-documents/promises.md

```diff
@@ -0,0 +1,112 @@
+### Why?
```

Comment:
After service separation we will have a lot of communication between services. Separating checkout adds > 20 additional requests. These brings two issues: increases loads (number of requests) and decreases performance (more communication). We will have to reduce amount of communication and optimize services themselves to minimize/avoid these issues.

Invoker in this proposal, from what I understand, will generate even more requests (`wait` method) which are not cheap. It will also increase execution time by much because we can send requests to get result of operation only with certain interval, which in our case can not be small. If we have composite operation, total time will be even bigger. The interval will be very hard to get right, as it should depend on load and average response time.

Because of the concerns above, I would like to understand better where where will we use this API and make sure it's good fit for these use cases.